File: leetcode/q2095/q2095.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Test Case 1: 1 Element
        if (head.next == None):
            return None

        # Test Case 2: 2 Elements
        if (head.next.next == None):
            head.next = None
            return head
        
        skip: bool = True
        mid: Optional[ListNode] = head
        end: Optional[ListMode] = head.next

        while end.next != None:
            if (skip == False):
                mid = mid.next
                skip = True
            else:
                skip = False
        
            end = end.next

        mid.next = mid.next.next
        return head

File: leetcode/q2095/q2095.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode deleteMiddle(ListNode head) {
        // Edge Case 1: 1 Element.
        if (head.next == null) {
            return null;
        }

        // Edge Case 2: 2 elements, remove element 2, return element 1.
        if (head.next.next == null) {
            head.next = null;
            return head;
        }

        ListNode half = head;
        ListNode end = head.next;
        boolean skip = true;

        while (end.next != null) {
            if (!skip) {
                half = half.next;
                skip = true;
            }
            else {
                skip = false;
            }
            end = end.next;
        }

        half.next = half.next.next;
        return head;
    }
}
